### PR TITLE
Individual training plans

### PR DESCRIPTION
**Pull request by wsbrenk**
1. changes proposed in this pull request:
 
   - fixes issue #587 and #623 

changes of branch 623 are included. Feel free to merge them one by one or in a single step.
 
 
2. changelog and release_notes ...

 - [x] have been updated
 - [ ] do not require update

3. [Optional] suggested person to review this PR @___

### CODE REVIEW

**Comment by akasolace** [comment]
@yaute74  I think you are the best person to do this review but if you have no time I will do it as I would like to have it merged before my holiday. Please let me know how you want to proceed.

**Comment by yaute74** [comment]
@akasolace I am sorry, but I will have  not the time actually to do a correct review

**Comment by akasolace** [comment]
> @akasolace I am sorry, but I will have not the time actually to do a correct review

ok no worry I will do it then ...

File: src/main/java/core/db/DbUtil.java [hunk 1]
```diff
@@ -5,10 +5,16 @@
...
 public class DbUtil {
 
-    static public Integer getNullableInt(ResultSet rs, String cl) throws SQLException {
+    public static Integer getNullableInt(ResultSet rs, String cl) throws SQLException {
         Integer res = rs.getInt(cl);
         if (!rs.wasNull()) return  res;
         return null;
     }
 
+    public static Boolean getNullableBoolean(ResultSet rs, String cl) throws SQLException {
```
Comment (akasolace) [inline src/main/java/core/db/DbUtil.java:14]: usually it is good practice to put the @nullabale operator to faciilitate the work of the debugger

File: src/main/java/core/model/match/MatchStatistics.java [hunk 2]
```diff
@@ -91,6 +105,7 @@ public int getTrainMinutesPlayedInPositions(int spielerId, int[] accepted) {
 		// Done with substitutions, add end if necessary
 
 		if (inPosition) {
+			// TODO: if opponent is walkover +=90
```
Comment (akasolace) [inline src/main/java/core/model/match/MatchStatistics.java:108]: is it a todo you intended to implement before submiting the PR ? do you want to open a dedicated ticket for this?

File: src/main/java/core/net/OnlineWorker.java [hunk 3]
```diff
@@ -266,6 +266,8 @@ public static boolean downloadMatchData(int matchid, MatchType matchType, boolea
 
 	public static boolean downloadMatchData( MatchKurzInfo info, boolean refresh)
 	{
+		if ( HOMainFrame.launching.get() ) return false; // do not call this on HO launch
```
Comment (akasolace) [inline src/main/java/core/net/OnlineWorker.java:269]: is this linked to the fix you made regarding skill of player who recently join the team? HO! tries to dtermine how many training the players received since last skill before joigning?  
also I haven't check but could we not simply avoid the function to be called during startup ?

File: src/main/java/core/training/TrainingManager.java [hunk 4]
```diff
@@ -179,79 +177,182 @@ public void recalcSubskills(boolean showBar) {
...
+					// Player appears the first time
+					// - was bought new
+					// - promoted from youth
+					// - it is the first hrf ever loaded
+					old = new Player();
+					old.setSpielerID(player.getSpielerID());
+					old.copySkills(player);
+					if (HOVerwaltung.instance().getModel().getCurrentPlayer(player.getSpielerID()) != null) {
+						// PLayer is in current team (not an historical player)
+						List<TrainingEvent> events = player.downloadTrainingEvents();
```
Comment (akasolace) [inline src/main/java/core/training/TrainingManager.java:212]: is this the line which is trigered at staturp and cause match download ?  with the modifictation to match downloader would that stillwork correctly ?

**Comment by akasolace** [comment]
> new database table futurePlayerTraining is ready. FutureTrainingsManager calculation is changed. Since there is no entry in the new table, the results of the training forecast are unchanged.  Now i need a good idea how to implement the editing of the individual future training plans. my very first idea was to add combo boxes to the cells of the training forecast overview table. can someone tell me if that's a stupid idea?

@wsbrenk we agree this is not supposed to be treated in this PR, you just prepare the table but there is still no logic to fill it or use it, right ?

**Comment by akasolace** [comment]
#587 seems closed by this PR for what I can tell, also it seems player have the right colors based on best positions (which in my case is very different to the actual trained positions so looking forward for the individual training plans as you don't want to use last training as a basis for the future as you explained in a different ticket )

**Comment by akasolace** [comment]
I can't check if #623 using my database but I think we could check this one down the road

**Comment by wsbrenk** [comment]
further answers tomorrow

**Comment by wsbrenk** [comment]
> 
> 
> > new database table futurePlayerTraining is ready. FutureTrainingsManager calculation is changed. Since there is no entry in the new table, the results of the training forecast are unchanged.  Now i need a good idea how to implement the editing of the individual future training plans. my very first idea was to add combo boxes to the cells of the training forecast overview table. can someone tell me if that's a stupid idea?
> 
> @wsbrenk we agree this is not supposed to be treated in this PR, you just prepare the table but there is still no logic to fill it or use it, right ?

in the meantime there the popup that fills this table and future training calculation uses this information.

**Comment by wsbrenk** [comment]
> 
> 
> #587 seems closed by this PR for what I can tell, also it seems player have the right colors based on best positions (which in my case is very different to the actual trained positions so looking forward for the individual training plans as you don't want to use last training as a basis for the future as you explained in a different ticket )


@akasolace are you able to select some cells (future training weeks)? If you release the mouse a popup should apear, that let's you select the training priority you want for the selected weeks.

File: src/main/java/core/db/DbUtil.java (hunk 1 above)
Comment (wsbrenk) [reply src/main/java/core/db/DbUtil.java:14]: public static @Nullable Integer getNullableInt(ResultSet rs, String cl) throws SQLException

@akasolace  do you mean it this way? i will change that

**Comment by wsbrenk** [comment]
> Also it is likely I missed it but I did not come across the db intiliaer. Just to be sure that users starting a fresh db will have the new training tables (i.e. same table as the ones going through dbUpdater)

to be honest: i didn't test this either. but i'm optimistic that introducing the table FuturePlayerTrainingTable in DBManager.initAllTables will do the job.

File: src/main/java/core/model/match/MatchStatistics.java (hunk 2 above)
Comment (wsbrenk) [reply src/main/java/core/model/match/MatchStatistics.java:108]: this line is left over from my first plan. finally, the treatment of walkover matches is done a few lines above. 
i will remove this line.

File: src/main/java/core/net/OnlineWorker.java (hunk 3 above)
Comment (wsbrenk) [reply src/main/java/core/net/OnlineWorker.java:269]: I completely agree with you - but I have no idea how this can be achieved. This is triggered by a change in MatchStatistic.getHomeTeamInfo, which I needed to avoid an npe. In my database I found matches with missing lineups ("Verlegenheitstruppe") that I wanted to repair by reloading.
Unfortunately, this method is also called at HO start.

File: src/main/java/core/training/TrainingManager.java (hunk 4 above)
Comment (wsbrenk) [reply src/main/java/core/training/TrainingManager.java:212]: see above - but i'm not sure if this method will also call the OnlineWorker

File: src/main/java/core/db/DbUtil.java (hunk 1 above)
Comment (akasolace) [reply src/main/java/core/db/DbUtil.java]: yes, and you will need to import org.jetbrains.annotations.Nullable;

File: src/main/java/core/model/match/MatchStatistics.java (hunk 2 above)
Comment (akasolace) [reply src/main/java/core/model/match/MatchStatistics.java]: ok

File: src/main/java/core/net/OnlineWorker.java (hunk 3 above)
Comment (akasolace) [reply src/main/java/core/net/OnlineWorker.java:269]: @wsbrenk ok maybe then you can add a todo explaining the issue for when a dev would have some time ... without explanation it will be quite confusing ....

File: src/main/java/core/training/TrainingManager.java (hunk 4 above)
Comment (akasolace) [reply src/main/java/core/training/TrainingManager.java:212]: ok let just put a todo as mentioned above and we can leave it as is for the time being ....

**Comment by akasolace** [comment]
> to be honest: i didn't test this either. but i'm optimistic that introducing the table FuturePlayerTrainingTable in DBManager.initAllTables will do the job.

If you put it in initAllTables  it should do the job ...

**Comment by akasolace** [comment]
> @akasolace are you able to select some cells (future training weeks)? If you release the mouse a popup should apear, that let's you select the training priority you want for the selected weeks.

I can select the training type but I am not sure this is whawt you are talking about ....

image

but I can 't find a way to select whic player recieve which kind of training (full/partial), is this doable ? As already mentioned in my case the detection based on best position is completly inaccurate and I think it will be the case for plenty of users ....

**Comment by wsbrenk** [comment]
@akasolace sorry, no you have to do it in the prediction panel:

grafik

**Comment by akasolace** [comment]
> @akasolace sorry, no you have to do it in the prediction panel:
> 
> 
> 
> grafik
> 
> 

Ah ok i will try again later

**Comment by akasolace** [comment]
> @akasolace sorry, no you have to do it in the prediction panel:
> 
> grafik

ah ok I found it no but I have some  remarks:
      - I think we should add this contextual menu to the Training tab to set the full / partial for the full period (the prediction panel could be used for fine tuning for a specific week but I think that usually users will want to set it for the full future)
      - in the prediction panel switching to full / partial / osmoesis works well but swithing to none is not working

**Comment by wsbrenk** [comment]
> - I think we should add this contextual menu to the Training tab to set the full / partial for the full period (the prediction panel could be used for fine tuning for a specific week but I think that usually users will want to set it for the full future)

the intention was to select training priorities for each player individually. it is not realistic/virtualistic to set full training for all of your players, because the number of available trainings positions will be limited.

> - in the prediction panel switching to full / partial / osmoesis works well but swithing to none is not working

that's a bit tricky. we have to distinguish between reset of the individual training setting for one player and setting NO training. in the moment the first is implemented and by resetting the individual training the automatic training depending on players best position is in place.

**Comment by wsbrenk** [comment]
> usually users will want to set it for the full future)

now i got your point. you want to select the priority in training panel for the whole time range by overriding the best position information via popup menu. that could be done i think. good idea!!!

**Comment by akasolace** [comment]
> > usually users will want to set it for the full future)
> 
> now i got your point. you want to select the priority in training panel for the whole time range by overriding the best position information via popup menu. that could be done i think. good idea!!!

yes it is exactly that ....

**Comment by wsbrenk** [comment]
this commit moves popup to an extra class, that i want to use in training panel too.

No training and resetting to best position training are separated items now.

File: src/main/java/core/net/OnlineWorker.java (hunk 3 above)
Comment (wsbrenk) [reply src/main/java/core/net/OnlineWorker.java]: @akasolace downloadMatchData is no longer called on HO launch. I removed this line.

File: src/main/java/core/training/TrainingManager.java (hunk 4 above)
Comment (wsbrenk) [reply src/main/java/core/training/TrainingManager.java:212]: @akasolace line is removed

**Comment by wsbrenk** [comment]
@akasolace now you can select training prio in training panel too:

grafik

File: build.gradle [hunk 5]
```diff
@@ -147,8 +147,7 @@ distributions {
 
 
 install4j {
-    if (install4jHomeDir == null) installDir = file("foo")
-    else installDir = file(install4jHomeDir)
+    if (hasProperty('install4jHomeDir'))  installDir = file(install4jHomeDir)
```
Comment (akasolace) [inline build.gradle:150]: I have to revert this change as it breaks my gradle task.
It seems install4j refuse to run if during the configuration phase variable installDir is not set this is why it has to be also specified in else clause

Comment (wsbrenk) [reply build.gradle:150]: > 
> 
> I have to revert this change as it breaks my gradle task.
> It seems install4j refuse to run if during the configuration phase variable installDir is not set this is why it has to be also specified in else clause

hi @akasolace : could you try it this way: 

    if (hasProperty('install4jHomeDir'))  installDir = file(install4jHomeDir)
    else installDir = file("foo")

this would work in my case.

Comment (akasolace) [reply build.gradle:150]: @wsbrenk arf not yet but the following should (hopefully) work for both of us :-)
```
    if ((hasProperty('install4jHomeDir')) || ((install4jHomeDir != null) )) installDir = file(install4jHomeDir)
    else installDir = file("foo")
```

Comment (wsbrenk) [reply build.gradle:150]: nope: A problem occurred evaluating root project 'HO'.
> path may not be null or empty string. path=''

what's about this on your side?

    if (hasProperty('install4jHomeDir')){
        if (install4jHomeDir != null) installDir = file(install4jHomeDir)
        else installDir = file("foo")
    }
